### ISSUE TITLE
Mosek dualization heuristic

### ISSUE THREAD

**Issue by phschiele** [cvxpy/cvxpy#2107]
As a result of a recent [discussion](https://github.com/cvxpy/cvxpy/discussions/2087) we identified the potential benefit of having heuristics to decide whether or not to dualize problem before handing them over to Mosek.

Citing some arguments by @aszekMosek from the discussion:
> Now a fair point would be, that the solver could itself decide whether to dualize if it pays off (at least heuristically). In fact MOSEK does it for linear and conic problems, but unfortunately not SDP. 

This suggests to me that we could disable the currently automatic dualization step for non-SDP problems?

> [For SDP problems] The deciding factor is how many constraints you will have that contain semidefinite terms.
We could start by building a simple heuristic based on the number of constraints similar to the [reference](https://docs.mosek.com/modeling-cookbook/duality.html#semidefinite-duality-and-lmis) provided in the discussion.

**Comment by SteveDiamond** [comment]
That's great @Paulnkk can look into this! Mosek internally has automatic dualization for non-SDP problems so I'm not sure there would be benefit for that case. The key issue is SDPs.

**Comment by phschiele** [comment]
@SteveDiamond Exactly, however, we currently dualize all non-integer problems:
https://github.com/cvxpy/cvxpy/blob/a6e88643570c3b4852d62d5932ba3db7c2b10bd9/cvxpy/reductions/solvers/conic_solvers/mosek_conif.py#L236-L239
Shouldn't we only apply the `Dualize` reduction only when it is an SDP then?

**Comment by SteveDiamond** [comment]
Agreed, we should avoid dualizing if there's no benefit because the verbose output is more intuitive without dualization. Currently the primal formulation doesn't support SDPs so need to add that in as well.

**Comment by rileyjmurray** [comment]
My motivation for adding it was not just SDP. My experience is that
dualization — purely as a means of eliminating slack variables — can reduce
the time to solve large exponential cone programs by orders of magnitude.

Just to emphasize again: we dualize to account for a mismatch in CVXPY’s
and MOSEK’s primal standard forms. The dualization avoids the introduction
of artificial slack variables during canonicalization.

**Comment by SteveDiamond** [comment]
@rileyjmurray would you suggest we replace the existing primal form code with the new mosek conic interface?

**Comment by rileyjmurray** [comment]
@SteveDiamond if I understand you correctly, "the existing primal form code" is the code for mixed-integer problems where we manually introduce slack variables. If that's the case then using the new Mosek conic interface would be equivalent and a bit simpler. We wouldn't want to replace our code outright because that would break things for users of Mosek 9 and lower.

**Comment by SteveDiamond** [comment]
Yeah, that's what I meant. Good point about keeping support for Mosek <= 9. Rather than adding the conic interface would it make sense then to add SDP support to the code path with slack variables and then add the automatic dualization layer on top, but use it for both SDP and other problems, given that it makes such a big difference for exponential cone problems?

**Comment by Paulnkk** [comment]
I would be very happy to work on that task! From how I understand it now, add the heuristic on SDPs based on the reference provided to decide whether to hand over Primal or dual form of SDP to Mosek? Concerning the other point with the slack variables; do I understand it correctly to apply this procedure also where slack variables are artificially introduced to potentially reduce runtime by comparing primal (with slacks) and dual ?
